core/queryparser/sparql/src/test/java/org/eclipse/rdf4j/query/parser/sparql/SPARQLParserTest.java

```diff
@@ -283,4 +287,34 @@ public void testLongUnicode() throws Exception {
 		assertEquals("\uD83D\uDE1F", lines[lines.length - 1].replaceAll(".*\"(.*)\".*", "$1"));
 	}
 
+	@Test
+	public void testWildCardPathFixedEnd() {
+
+		String query = "PREFIX : <http://example.org/>\n ASK {:IBM ((:|!:)|(^:|!^:))* :Jane.} ";
+
+		ParsedQuery parsedQuery = parser.parseQuery(query, null);
+
+		// parsing should not throw exception
+		TupleExpr tupleExpr = parsedQuery.getTupleExpr();
+
+	}
+
+	@Test
+	public void testWildCardPathOpenEnd() {
+
+		String query = "PREFIX : <http://example.org/>\n ASK {:IBM ^(:|!:) ?jane.} ";
+
+		ParsedQuery parsedQuery = parser.parseQuery(query, null);
+		TupleExpr tupleExpr = parsedQuery.getTupleExpr();
+
+		Slice slice = (Slice) tupleExpr;
+		Union union = (Union) slice.getArg();
+
+		Var leftSubjectVar = ((StatementPattern) union.getLeftArg()).getSubjectVar();
+		Var rightSubjectVar = ((StatementPattern) ((Filter) union.getRightArg()).getArg()).getSubjectVar();
+
+		assertEquals(leftSubjectVar, rightSubjectVar);
+
```

Comment:
Going to need help with fixing the underlying issue for this test.

Reply:
Yes, unfortunately this code is the worst kind of god-class. I think I see what needs to be fixed, but I'll need a bit of time.